embassy-mspm0/src/i2c.rs

```diff
@@ -0,0 +1,1156 @@
+pub const FIFO_SIZE: usize = 8;
+
+/// I2C Driver.
+pub struct I2c<'d, M: Mode> {
+    info: &'static Info,
+    state: &'static State,
+    scl: Option<Peri<'d, AnyPin>>,
+    sda: Option<Peri<'d, AnyPin>>,
+    _phantom: PhantomData<M>,
+}
+
+impl<'d, M: Mode> SetConfig for I2c<'d, M> {
+    type Config = Config;
+    type ConfigError = ConfigError;
+
+    fn set_config(&mut self, config: &Self::Config) -> Result<(), Self::ConfigError> {
+        self.set_config(config)
+    }
+}
+
+impl<'d> I2c<'d, Blocking> {
+    pub fn new_blocking<T: Instance>(
+        peri: Peri<'d, T>,
+        scl: Peri<'d, impl SclPin<T>>,
+        sda: Peri<'d, impl SdaPin<T>>,
+        config: Config,
+    ) -> Result<Self, ConfigError> {
+        if !config.check_clock_rate() {
+            return Err(ConfigError::InvalidClockRate);
+        }
+
+        Self::new_inner(peri, scl, sda, config)
+    }
+}
+
+impl<'d> I2c<'d, Async> {
+    pub fn new_async<T: Instance>(
+        peri: Peri<'d, T>,
+        scl: Peri<'d, impl SclPin<T>>,
+        sda: Peri<'d, impl SdaPin<T>>,
+        _irq: impl Binding<T::Interrupt, InterruptHandler<T>> + 'd,
+        config: Config,
+    ) -> Result<Self, ConfigError> {
+        if !config.check_clock_rate() {
+            return Err(ConfigError::InvalidClockRate);
+        }
+
+        let i2c = Self::new_inner(peri, scl, sda, config);
+
+        T::info().interrupt.unpend();
+        unsafe { T::info().interrupt.enable() };
+
+        i2c
+    }
+}
+
+impl<'d, M: Mode> I2c<'d, M> {
+    /// Reconfigure the driver
+    pub fn set_config(&mut self, config: &Config) -> Result<(), ConfigError> {
+        if !config.check_clock_rate() {
+            return Err(ConfigError::InvalidClockRate);
+        }
+
+        self.info.interrupt.disable();
+
+        if let Some(ref sda) = self.sda {
+            sda.update_pf(config.sda_pf());
+        }
+
+        if let Some(ref scl) = self.scl {
+            scl.update_pf(config.scl_pf());
+        }
+
+        self.init(config)
+    }
+
+    fn init(&mut self, config: &Config) -> Result<(), ConfigError> {
+        // Init I2C
+        self.info.regs.clksel().write(|w| match config.clock_source {
+            ClockSel::BusClk => {
+                w.set_mfclk_sel(false);
+                w.set_busclk_sel(true);
+            }
+            ClockSel::MfClk => {
+                w.set_mfclk_sel(true);
+                w.set_busclk_sel(false);
+            }
+        });
+        self.info.regs.clkdiv().write(|w| w.set_ratio(config.clock_div.into()));
+
+        // set up glitch filter
+        self.info.regs.gfctl().modify(|w| {
+            w.set_agfen(false);
+            w.set_agfsel(vals::Agfsel::AGLIT_50);
+            w.set_chain(true);
+        });
+
+        // Reset controller transfer, follow TI example
+        self.info
+            .regs
+            .controller(0)
+            .cctr()
+            .write_value(i2c::regs::Cctr::default());
```

Comment:
Do reserved bits need to be kept at boot values?

Reply:
thanks, fixed